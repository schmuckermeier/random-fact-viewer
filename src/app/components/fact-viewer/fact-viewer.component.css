.button-list {
  button {
    margin-right: 2rem;
    margin-top: 1rem;
  }

  @media (max-width: var(--mobile-breakpoint)) {
    text-align: center;
  }
}
