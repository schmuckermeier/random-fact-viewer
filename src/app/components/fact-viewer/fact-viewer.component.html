<mat-card appearance="outlined">
  <mat-card-content>{{ currentRandomFact()?.text }}</mat-card-content>
</mat-card>
<div class="button-list">
  <button extended mat-fab (click)="fetchRandomFact()">
    <mat-icon>refresh</mat-icon>
    Load a random fact
  </button>
  <button extended mat-fab (click)="saveAsFavorite(currentRandomFact())" [disabled]="!currentRandomFact()">
    <mat-icon>favorite</mat-icon>
    Save as favorite
  </button>
</div>
