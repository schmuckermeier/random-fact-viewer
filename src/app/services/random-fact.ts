export type RandomFact = {
  id: string;
  "text": string,
  "source": string,
  "source_url": string,
  "language": "en"| "de",
  "permalink": string
}
