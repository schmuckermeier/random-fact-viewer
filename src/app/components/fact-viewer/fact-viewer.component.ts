import {Component, inject, signal} from '@angular/core';
import {RandomFact} from '../../services/random-fact';
import {RandomFactService} from '../../services/random-fact.service';
import {FavoriteFactsService} from '../../services/favorite-facts.service';
import {take} from 'rxjs';
import {MatCardModule} from '@angular/material/card';
import {MatIconModule} from '@angular/material/icon';
import {MatButtonModule} from '@angular/material/button';

@Component({
  selector: 'app-fact-viewer',
  imports: [MatCardModule,MatIconModule, MatButtonModule],
  templateUrl: './fact-viewer.component.html',
  styleUrl: './fact-viewer.component.css'
})
export class FactViewerComponent {
  private randomFactService = inject(RandomFactService);
  private favoriteFactsService = inject(FavoriteFactsService);

  currentRandomFact = signal<RandomFact | undefined>(undefined);

  constructor() {
    this.fetchRandomFact();
  }

  fetchRandomFact() {
    this.randomFactService.getRandomFact().pipe(take(1)).subscribe(randomFact => {
      this.currentRandomFact.set(randomFact)
    })
  }

  saveAsFavorite(fact: RandomFact | undefined) {
    fact && this.favoriteFactsService.saveFact(fact)
  }
}
