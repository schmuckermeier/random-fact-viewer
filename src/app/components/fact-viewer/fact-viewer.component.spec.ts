import {ComponentFixture, TestBed} from '@angular/core/testing';
import {FactViewerComponent} from './fact-viewer.component';
import {RandomFactService} from '../../services/random-fact.service';
import {mockRandomFact, mockRandomFacts} from '../../testing/mockRandomFacts';
import {of} from 'rxjs';
import {FavoriteFactsService} from '../../services/favorite-facts.service';
import {queryById} from '../../testing/testing-utils';

describe('FactViewerComponent', () => {
  let component: FactViewerComponent;
  let fixture: ComponentFixture<FactViewerComponent>;
  let favoriteFactsService: FavoriteFactsService;

  let mockRandomFactService = {
    getRandomFact: jest.fn(() => of(mockRandomFact))
  };

  let mockFavoriteFactsService = {
    saveFact: jest.fn()
  };

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [FactViewerComponent],
      providers: [
        {provide: FavoriteFactsService, useValue: mockFavoriteFactsService},
        {provide: RandomFactService, useValue: mockRandomFactService}]
    })
      .compileComponents();

    fixture = TestBed.createComponent(FactViewerComponent);
    component = fixture.componentInstance;
    favoriteFactsService = TestBed.inject(FavoriteFactsService);

    fixture.detectChanges();
  });

  it('should render with initial fact and buttons', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should load first random fact', () => {
    expect(component.currentRandomFact()).toEqual(mockRandomFact);
  });

  it('should load next random fact on click', () => {
    queryById(fixture, "loadFact").nativeElement.click();
    expect(mockRandomFactService.getRandomFact).toHaveBeenCalled();
  });

  it('should save random fact on click', () => {
    queryById(fixture, "saveFact").nativeElement.click();
    expect(favoriteFactsService.saveFact).toHaveBeenCalledWith(mockRandomFact);
  });

  it('should show selected fact if it was set', () => {
    const nextRandomFact = mockRandomFacts[1];
    component.selectedFact = nextRandomFact;
    fixture.detectChanges();
    const viewerContent = queryById(fixture, "viewerContent").nativeElement.innerHTML;
    expect(viewerContent).toEqual(nextRandomFact.text);
  });
});
