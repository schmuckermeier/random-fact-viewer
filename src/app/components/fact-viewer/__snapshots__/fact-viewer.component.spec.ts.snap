// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FactViewerComponent should render with initial fact and buttons 1`] = `
<app-fact-viewer
  currentRandomFact={[Function Function]}
  favoriteFactsService={[Function Object]}
  randomFactService={[Function Object]}
>
  <mat-card
    appearance="outlined"
    class="mat-mdc-card mdc-card mat-mdc-card-outlined mdc-card--outlined"
  >
    <mat-card-content
      class="mat-mdc-card-content"
      id="viewerContent"
    >
      Golf courses cover 4% of North America.
    </mat-card-content>
  </mat-card><div
    class="button-list"
  >
    <button
      class="mdc-fab mat-mdc-fab-base mat-mdc-fab mat-accent mat-mdc-button-base mdc-fab--extended mat-mdc-extended-fab"
      extended=""
      id="loadFact"
      mat-fab=""
      mat-ripple-loader-class-name="mat-mdc-button-ripple"
      mat-ripple-loader-uninitialized=""
    >
      <span
        class="mat-mdc-button-persistent-ripple mdc-fab__ripple"
      />
      <mat-icon
        aria-hidden="true"
        class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
        data-mat-icon-type="font"
        role="img"
      >
        refresh
      </mat-icon>
      <span
        class="mdc-button__label"
      >
         Load a random fact 
      </span>
      <span
        class="mat-focus-indicator"
      />
      <span
        class="mat-mdc-button-touch-target"
      />
    </button>
    <button
      class="mdc-fab mat-mdc-fab-base mat-mdc-fab mat-accent mat-mdc-button-base mdc-fab--extended mat-mdc-extended-fab"
      extended=""
      id="saveFact"
      mat-fab=""
      mat-ripple-loader-class-name="mat-mdc-button-ripple"
      mat-ripple-loader-uninitialized=""
    >
      <span
        class="mat-mdc-button-persistent-ripple mdc-fab__ripple"
      />
      <mat-icon
        aria-hidden="true"
        class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
        data-mat-icon-type="font"
        role="img"
      >
        favorite
      </mat-icon>
      <span
        class="mdc-button__label"
      >
         Save as favorite 
      </span>
      <span
        class="mat-focus-indicator"
      />
      <span
        class="mat-mdc-button-touch-target"
      />
    </button>
  </div>
</app-fact-viewer>
`;
